### PR TITLE
atc/wrappa: wrappa to reject archived pipelines

### LINKED ISSUE #5430: api/fly: write operations to an archived pipeline should fail
**Epic:** #5055

Generally speaking, given a pipeline is archived, the majority of write operations done via the API should fail. Below goes in full detail.

**Given**
```
$ fly -t dev set-pipeline -p test -c topgun/pipelines/custom-types.yml
$ fly -t dev archive-pipeline -p test
```

**Then**
* [x] API action `GetConfig` will `404`; when a pipeline is archived its config is removed to avoid leaking sensitive information.
  * Currently:
    ```
    HTTP/1.1 200 OK
		{"config":{"jobs":[{"name":"","plan":null}]}}
    ```
  * Desired:
    ```
    $ fly -t dev get-pipeline -p test
    <some error>
    ```
* [x] API action `CreateJobBuild` will error; archived pipelines must consume no scheduling resources, not even the build starter.

  * Currently:
    ```
    CreateJobBuild

    POST /api/v1/teams/main/pipelines/archived-2/jobs/test/builds HTTP/1.1
    2020/04/13 12:22:01 HTTP/1.1 200 OK
    ```
  * Desired:
    ```
    $ fly -t dev trigger-job -j test/job
    <some error>
    ```
* [x] API action `ScheduleJob` will error; archived pipelines must consume no scheduling resources, not even the build starter.

  * Currently:
    ```
    PUT /api/v1/teams/main/pipelines/archived-2/jobs/test/schedule HTTP/1.1
    HTTP/1.1 200 OK
    ```
  * Desired:
    ```
    $ fly -t dev schedule-job -j test/job
    <some error>
    ```
* [x] `CheckResource` and `CheckResourceType` will error; archived pipelines must consume no scheduling resources, including queuing checks.

  * Currently:
    ```
    CheckResource

		POST /api/v1/teams/main/pipelines/archived-2/resources/interval/check HTTP/1.1
    {"from":null}
		HTTP/1.1 201 Created

		CheckResourceType

    POST /api/v1/teams/main/pipelines/archived-2/resource-types/time/check HTTP/1.1
    {"from":null}
    HTTP/1.1 201 Created

    ```
  * Desired:
    ```
    $ fly -t dev check-resource -r test/resource-name
    <some error>
    $ fly -t dev check-resource-type -r test/resource-name
    <some error>
    ```
* [x] `PinResourceVersion`, `UnpinResource`, `EnableResourceVersion`, `DisableResourceVersion`, and `SetPinCommentOnResource` will error; archived pipelines are read-only.
	all should probably be revisited:
  * Currently (only tested one, assumed the rest behave the same):
    ```
    PinResourceVersion

    PUT /api/v1/teams/main/pipelines/archived-2/resources/interval/versions/3/pin
    HTTP/1.1 200 OK
    ```
  * Desired:
    ```
    $ fly -t dev pin-resource -r test/resource-name -v version --comment "some comment"
    <some error>
    $ fly -t dev unpin-resource -r test/resource-name
    <some error>
    $ fly -t dev enable-resource-version -r test/resource-name -v version
    <some error>
    $ fly -t dev disable-resource-version -r test/resource-name -v version
    <some error>
    ```

### PR DESCRIPTION
**Pull request by taylorsilva**
# What is this PR trying to accomplish?

closes #5430

# How does it accomplish that?

Any endpoints that are wrapped will return a HTTP 409 conflict error if the pipeline is archived.

Like the api_auth wrappa, this wrappa will panic for any unhandled API
endpoints.

# Contributor Checklist

- [x] Unit tests
- [ ] Integration tests
- [ ] Updated documentation (located at https://github.com/concourse/docs)
- [ ] Updated release notes (located at https://github.com/concourse/concourse/tree/master/release-notes)

# Reviewer Checklist

- [ ] Code reviewed
- [ ] Tests reviewed
- [ ] Documentation reviewed
- [ ] Release notes reviewed
- [ ] PR acceptance performed
- [ ] New config flags added? Ensure that they are added to the [BOSH](https://github.com/concourse/concourse-bosh-release) 
      and [Helm](https://github.com/concourse/helm) packaging; otherwise, ignored for the [integration tests](https://github.com/concourse/ci/tree/master/tasks/scripts/check-distribution-env) (for example, if they are Garden configs that are not displayed in the `--help` text).

### CODE REVIEW

File: atc/wrappa/reject_archived_wrappa_test.go [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+			atc.CheckResource:           raHandlerFactory.RejectArchived(inputHandlers[atc.CheckResource]),
+			atc.CheckResourceType:       raHandlerFactory.RejectArchived(inputHandlers[atc.CheckResourceType]),
+			atc.DisableResourceVersion:  raHandlerFactory.RejectArchived(inputHandlers[atc.DisableResourceVersion]),
+			atc.EnableResourceVersion:   raHandlerFactory.RejectArchived(inputHandlers[atc.EnableResourceVersion]),
+			atc.PinResourceVersion:      raHandlerFactory.RejectArchived(inputHandlers[atc.PinResourceVersion]),
+			atc.UnpinResource:           raHandlerFactory.RejectArchived(inputHandlers[atc.UnpinResource]),
+			atc.SetPinCommentOnResource: raHandlerFactory.RejectArchived(inputHandlers[atc.SetPinCommentOnResource]),
+			atc.RerunJobBuild:           raHandlerFactory.RejectArchived(inputHandlers[atc.RerunJobBuild]),
+
+			atc.GetConfig:                     inputHandlers[atc.GetConfig],
```
Comment (cirocosta) [inline atc/wrappa/reject_archived_wrappa_test.go:46]: I wonder if there's a way of getting having this map explicitly filled only with those that are `RejectArchived` and the rest coming from a `for ...` :thinking:

File: atc/wrappa/reject_archived_wrappa.go [hunk 2]
```diff
@@ -0,0 +1,190 @@
...
+			atc.UnpausePipeline,
+			atc.ExposePipeline,
+			atc.HidePipeline,
+			atc.CreatePipelineBuild,
+			atc.ClearTaskCache,
+			atc.CreateArtifact,
+			atc.GetArtifact:
+
+		default:
+			panic("how do archived pipelines affect your endpoint?")
```
Comment (cirocosta) [inline atc/wrappa/reject_archived_wrappa.go:128]: noice :ok_hand:

File: atc/wrappa/reject_archived_wrappa_test.go (hunk 1 above)
Comment (taylorsilva) [reply atc/wrappa/reject_archived_wrappa_test.go:46]: Wouldn't that just move this list somewhere else? I'm trying to think how that would work along with the `panic` in the `default:` case...

Comment (aoldershaw) [reply atc/wrappa/reject_archived_wrappa_test.go:46]: Could you do something like this?

```go
inputHandlers := rata.Handlers{}
for _, route := range atc.Routes {
  inputHandlers[route.Name] = &stupidHandler{}
}

rejectArchiveRoutes := []string{
  atc.PausePipeline,
  atc.CreateJobBuild,
  ...
}
rejectArchiveLookup := make(map[string]bool)
for _, name := range rejectArchiveRoutes {
  rejectArchiveLookup[name] = true
}

wrappedHandlers := raWrappa.Wrap(inputHandlers)

for name, handler := range inputHandlers {
  expectedHandler := handler
  if rejectArchiveLookup[name] {
    expectedHandler = raHandlerFactory.RejectArchived(expectedHandler)
  }
  Expect(wrappedHandlers[name]).To(BeIdenticalTo(expectedHandler), "handler is "+name)
}
```

Comment (taylorsilva) [reply atc/wrappa/reject_archived_wrappa_test.go:46]: oh yeah, dumb me didn't realize this was the test file :face_palm:

**Comment by taylorsilva** [comment]
Ready for review!
The latest commit moves the handler and handler factory of `RejectArchived` to
the `pipelineserver` package. A test suite was also added to test the
http handler.

File: atc/api/pipelineserver/reject_archived_handler_factory.go [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+	delegateHandler http.Handler
+}
+
+func (ra RejectArchivedHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	teamName := r.FormValue(":team_name")
+	pipelineName := r.FormValue(":pipeline_name")
+
+	team, found, err := ra.teamFactory.FindTeam(teamName)
+	if err != nil {
+		w.WriteHeader(http.StatusInternalServerError)
```
Comment (cirocosta) [inline atc/api/pipelineserver/reject_archived_handler_factory.go:37]: nit: I think this error scenario is not covered :thinking:  aside from ` 	Context("when a team is not found", func() {`, we'd then have a "when finding a team fails" I guess

File: atc/api/pipelineserver/reject_archived_handler_factory.go [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+	}
+
+	if !found {
+		w.WriteHeader(http.StatusNotFound)
+		return
+	}
+
+	pipeline, found, err := team.Pipeline(pipelineName)
+	if err != nil {
+		w.WriteHeader(http.StatusInternalServerError)
```
Comment (cirocosta) [inline atc/api/pipelineserver/reject_archived_handler_factory.go:48]: nit: this seems to not be covered either

File: atc/api/pipelineserver/reject_archived_handler_factory_test.go [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+		fakePipeline = new(dbfakes.FakePipeline)
+
+		handlerFactory := pipelineserver.NewRejectArchivedHandlerFactory(dbTeamFactory)
+		handler = handlerFactory.RejectArchived(delegate.GetHandler(fakePipeline))
+	})
+
+	JustBeforeEach(func() {
+		server = httptest.NewServer(handler)
+
+		request, err := http.NewRequest("POST", server.URL+"?:team_name=some-team&:pipeline_name=some-pipeline", nil)
```
Comment (cirocosta) [inline atc/api/pipelineserver/reject_archived_handler_factory_test.go:42]: ha, neat!

File: atc/wrappa/reject_archived_wrappa.go [hunk 6]
```diff
@@ -0,0 +1,133 @@
...
```
Comment (cirocosta) [inline atc/wrappa/reject_archived_wrappa.go:126]: do you think we could test this? :thinking:  

e.g., using http://onsi.github.io/gomega/#panic

File: atc/api/pipelineserver/reject_archived_handler_factory_test.go (hunk 5 above)
Comment (taylorsilva) [reply atc/api/pipelineserver/reject_archived_handler_factory_test.go:42]: Thought so too! I copied it from the scoped handler test in the same package 😅

File: atc/wrappa/reject_archived_wrappa.go [hunk 7]
```diff
@@ -0,0 +1,133 @@
...
+			atc.GetCC,
+			atc.GetVersionsDB,
+			atc.ListJobInputs,
+			atc.OrderPipelines,
+			atc.PauseJob,
+			atc.ArchivePipeline,
+			atc.RenamePipeline,
+			atc.SaveConfig,
+			atc.UnpauseJob,
+			atc.UnpausePipeline,
```
Comment (aoldershaw) [inline atc/wrappa/reject_archived_wrappa.go:117]: Question: do we want to deny unpausing pipelines as well? Right now, we're doing this at the DB layer:

https://github.com/concourse/concourse/blob/b89d468b4ee86841f970d75433c2ae7ea3461124/atc/db/pipeline.go#L603-L605

Which we also check at the API level:

https://github.com/concourse/concourse/blob/3ac5db508adb9db262f52857d6d23ea2c92f6c7e/atc/api/pipelineserver/unpause.go#L14-L18

Might make sense to just use this wrappa

Comment (taylorsilva) [reply atc/wrappa/reject_archived_wrappa.go:117]: Good one, forgot about that one!

**Comment by taylorsilva** [comment]
@cirocosta can you review again please. I added Aidan's changes.

**Comment by cirocosta** [comment]
aah I've got a failure :thinking:  might be reported by `unit` soon:

```

------------------------------
• Failure [0.001 seconds]
Pipelines API
/home/ubuntu/workspace/concourse/atc/api/pipelines_test.go:21
  PUT /api/v1/teams/:team_name/pipelines/:pipeline_name/unpause
  /home/ubuntu/workspace/concourse/atc/api/pipelines_test.go:991
    when authenticated
    /home/ubuntu/workspace/concourse/atc/api/pipelines_test.go:1004
      when requester belongs to the team
      /home/ubuntu/workspace/concourse/atc/api/pipelines_test.go:1008
        when unpausing the pipeline fails with a conflict
        /home/ubuntu/workspace/concourse/atc/api/pipelines_test.go:1041
          returns 409 [It]
          /home/ubuntu/workspace/concourse/atc/api/pipelines_test.go:1049

          Expected
              <int>: 500
          to equal
              <int>: 409

          /home/ubuntu/workspace/concourse/atc/api/pipelines_test.go:1050
```

**Comment by cirocosta** [comment]
aaah seems like there's another failure :sweat_smile:  

```


• Failure [23.769 seconds]

ATC Integration Test [It] fails when unpausing an archived pipeline 

/tmp/build/0df9eea0/concourse/atc/integration/archiving_test.go:39


  Expected

      <string>: Unexpected Response

      Status: 409 Conflict

      Body:

      action not allowed for an archived pipeline

      

  to contain substring

      <string>: archived pipelines cannot be unpaused
```

https://ci.concourse-ci.org/builds/1148597#L5eba6a6a:150:162

also, wdyt of 

```

db/conflict.go
5://go:generate counterfeiter . Conflict
7:type Conflict interface {
9:      Conflict() string
12:type conflict string
14:func (c conflict) Error() string {
15:     return fmt.Sprintf("conflict error: %s", string(c))
18:func (c conflict) Conflict() string {
```

?

is that dead code now?

**Comment by taylorsilva** [comment]
From my last commit message:

> Made a small chagnge to the archiving integration test as well. Not 100%
sure if we should keep it or add more tests like it. I'm leaning towards
adding more tests like it for each of the endpoints that get wrapped in
the RejectArchived wrappa. If we decide to do that it will be in another
commit.

I'm still on the fence about this a bit. wdyt? @cirocosta 
1. keep this one test
2. add more tests like this for each endpoint (11 more)
3. delete the test since wrappa mostly covers this (server-side, no client-side coverage)
🤔

**Comment by cirocosta** [comment]
from my understanding, we have tests that verify:

- the behavior of what `wrappa` does
- the behavior of the handler itself

but we lack a test that ensure that `wrappa` is part of the middleware chain
that wraps our API requests.

it seems to me that if we get a single test that just verifies that, we can then
avoid having to duplicate all of the rest :thinking: wdyt?

**Comment by taylorsilva** [comment]
Yup, in that case the existing test, the last one that was failing in the integration test suite, covers that last case. Guess this PR is good to go!

**Comment by cirocosta** [comment]
cool! if you think it's redundant to keep all of those integration tests, might be good to reduce that in a different PR then :+1:

**Comment by taylorsilva** [comment]
@cirocosta I rebased on master which made your review stale. Was trying to get watsjs to pass!!! Could you approve again?

**Pull request merged**